docs/news/posts/2024-10-18-exploring-aiida-database.md

```diff
@@ -0,0 +1,339 @@
+    Out[10]: <StructureData: uuid: 9e65c339-2428-4fb1-8d75-a54da9f87879 (pk: 43956)>
+    ```
+
+    Note that we don't really need all of the intermediate node variables.
+    We did it mostly for convenience, but all of these steps can just be concatenated in a single long chain, as follows:
+
+    ```python
+    In [11]: bandgap_node.creator.inputs.bands.creator.inputs.structure
+    Out[11]: <StructureData: uuid: 9e65c339-2428-4fb1-8d75-a54da9f87879 (pk: 43956)>
+    ```
+
+    The advantage of the long string above is that after every dot you can use tab completion, and therefore (once you get used to it) it becomes very quick to browse advanced provenance graphs in the verdi shell.
+
+    One more thing one might want to do is to check if there is a better way to distinguish the `CalcFunctionNode` that I got at `Out[5]` above (stored in `calculation_node = bandgap_node.creator`).
+    Let us check its attributes:
+
+    ```python
+    In [12]: bandgap_node.creator.attributes.keys()
+    Out[12]: dict_keys(['first_line_source_code', 'source_code', 'source_file', 'function_name', 'sealed'])
+    In [13]: bandgap_node.creator.attributes['function_name']
+    Out[13]: 'get_bandgap_inline'
+    ```
+
+    This information will be useful in the following section, and will basically allow us to perform a query by filtering by the original function name.
+
+Now, after all these steps, we have a better understanding of the the structure of the data.
+Another useful tool to get a good idea of the connectivity is the graph generator.
+One can use `verdi node graph generate` to visualize the provenance surrounding a node.
+Limiting it to four levels up (ancestors) will be enough for this case; we will also avoid showing any descendants, since we are not interested here in calculations that used the data node as an input.
+Note that this command has to be executed outside of the verdi shell.
+
+```bash
+$ verdi node graph generate --process-in --process-out --ancestor-depth=4 --descendant-depth=0 138d1979-4aca-4a23-af18-38c707707585
+```
+
+The result should look something like this:
+
+![provenance](../pics/2024-exploring-aiida-database.png)
+
+## Systematic querying of the database
+
+Now that we have understood broadly the data layout (provenance links between nodes, their types, and some of the relevant attributes keys and values), let's now construct a query using the QueryBuilder in order to get the band structure of a set of 2D materials.
+Create a new text file and copy the content below (these are essentially python scripts, so you can use the `.py` extension).
+There are some comments which explain the purpose of each line of code, please read them carefully to understand how the query is constructed.
+If you are not familiar with the `QueryBuilder`, we refer to the [official AiiDA documentation for more details](https://aiida.readthedocs.io/projects/aiida-core/en/stable/howto/query.html).
+
+```python
```

Comment:
This big code block suddenly is quite a shocker. Should we provide a downloadable link to the python file instead? And then show the individual `append` calls in separate code snippets, explaining them one by one as I did in my last blog post about the query builder (at least the first two, and possibly less verbose than what I did in the past)?

For me, if I was reading this blog post, and suddenly this block of code would appear, I'd probably close the tab 😅 it also looks more intimidating because of the long comments. These are very useful, so I'd keep them in the `.py` file for sure, but here, like this, the code snippet basically fills the entire page, and people would probably spend 15 minutes understanding the snippet.

Reply:
I see your point, but I respectfully disagree. I think this script is one of the most valuable things for users, as they might want to come back to it and compare with their own queries. We could split it into snippets, but the length would be the same (or longer) and i think it would be more difficult to grasp. But i agree that the comments could be made a bit shorter, will do it. But we can also discuss further on this.